### PR TITLE
Code review changes

### PR DESCRIPTION
**Pull request by alandipert**
This incorporates feedback from the work week code review, a streamlined upload-recording mechanism, and changes the recording format slightly.

* `REQ` type requests, which indicate plain-HTTP with no special shiny protocol significance, are now known as `REQ_GET`. They correspond to `REQ_POST`. The `method` field has been removed in favor of this "top-level" approach to type differentiation.
* The `statusCode` field is now called `status` on `REQ_GET` and `REQ_POST` records
* `favicon.ico` requests are not recorded, and the SockJS `o` messages are also no longer recorded because their timing information is not interesting.
* `REQ_POST` requests record POSTs. If the POST has a body, its contents are recorded as `<logfile>.post.<n>` where `<logfile>` is the `output_file` argument passed to `record_session` and `n` is a number that increases by 1 every upload, starting from 0. POST bodies are stored to a file and sent upstream in chunks inside curl's `readfunction` callback.
* `REQ_{POST,GET}` requests now both include `begin` and `end` timestamps that correspond to the times before and after the request was completed. The `created` timestamp has been removed. `WS_*` type records include only a `begin` timestamp, as the completion of a message has no meaning at the websocket level - the interval between websocket messages does.
* A new `WS_RECV_BEGIN_UPLOAD` record containing the upload URL and ID as generated by the server was introduced. It contributes the `${UPLOAD_URL}` and `${UPLOAD_JOB_ID}` tokens to the session's token store.
* The token store is now an environment, and tokens are used exclusively to store replaceable strings associated with the session like worker id, sockjs identifiers, etc. Previously, private fields of the `RecordingSession` class were used to store these.
* A new `URLBuilder` class was introduced to encapsulate the logic of forming new URLs based on old ones.
* Cookies are now maintained between requests, which permits support for clustered RSC instances behind an ALB, as the ALB stickiness cookie is now retained and attached to the initial websocket request.

A corresponding set of adaptations will need to be made to shinycannon.

### CODE REVIEW

File: R/auth.R [hunk 1]
```diff
@@ -9,7 +9,11 @@ getInputs <- function(html, server) {
 }
 
 pasteParams <- function(df, collapse) {
-  paste0(df[["name"]], "=", df[["value"]], collapse = collapse)
+  if (nrow(df) == 0) {
+    ""
+  } else {
+    paste0(df[["name"]], "=", df[["value"]], collapse = collapse)
```
Comment (jcheng5) [inline R/auth.R]: No escaping needed for the name/value?

File: R/auth.R [hunk 2]
```diff
@@ -51,26 +55,23 @@ loginUrlFor <- function(appUrl, appServer) {
...
 handlePost <- function(handle, loginUrl, postfields, cookies, cookieName) {
   curl::handle_setopt(handle,
     postfields = postfields,
-    # TODO how should the cookies be encoding?
     cookie = pasteParams(cookies, "; "),
     post = TRUE,
     followlocation = FALSE,
-    ssl_verifyhost = 0, ssl_verifypeer = 0
+    ssl_verifyhost = 0,
+    ssl_verifypeer = 0
```
Comment (jcheng5) [inline R/auth.R]: Do you think ignoring SSL errors should be the default, or opt-in? (I could probably go either way, but ideally we wouldn't change it to be stricter once it's public)

File: R/shiny-recorder.R [hunk 3]
```diff
@@ -209,113 +166,164 @@ RecordingSession <- R6::R6Class("RecordingSession",
...
+
+      event <- list(
+        type = "REQ_POST",
+        begin = makeTimestamp(begin),
+        end = makeTimestamp(end),
+        status = resp_curl$status_code,
+        url = replaceTokens(paste0(req$PATH_INFO, req$QUERY_STRING), self$tokens)
+      )
+
+      event <- append(event, if (!is.null(dataFileName)) list(datafile = basename(dataFileName)))
```
Comment (jcheng5) [inline R/shiny-recorder.R]: I would just do
```
if (!is.null(dataFileName)) {
  event$datafile <- basename(dataFileName)
}
```

**Comment by jcheng5** [comment]
Also look at "hop by hop" http proxy headers

File: R/shiny-recorder.R [hunk 4]
```diff
@@ -67,90 +43,77 @@ getWorkerId <- function(page) {
...
-# Generate a new message based on originalMessage but with new contents
-# {"type":"WS_RECV_INIT","created":"2018-03-12T19:51:59.675Z","message":"a[\"1#0|m|{\\\\\"config\\\\\":{\\\\\"workerId\\\\\":[\\\\\"${WORKER}\\\\\"],\\\\\"sessionId\\\\\":[\\\\\"${SESSION}\\\\\"],\\\\\"user\\\\\":null}}\"]"}
-# {"type":"WS_RECV_INIT","created":"2018-03-12T20:21:27.086Z","message":"a[\"1#0|m|[\"{\\\"config\\\":{\\\"workerId\\\":[\\\"${WORKER}\\\"],\\\"sessionId\\\":[\\\"${SESSION}\\\"],\\\"user\\\":null}}\"]\"]"}
-spliceMessage <- function(originalMessage, newMessageObject) {
-  newMsg <- jsonlite::toJSON(jsonlite::unbox(jsonlite::toJSON(newMessageObject, null = 'null')))
-  group <- stringr::str_match(originalMessage, messagePattern)
-  paste0(group[1,2], group[1,3], group[1,4], newMsg, group[1,6])
+replaceTokens <- function(str, tokens) {
+  if (length(tokens) > 0) {
+    stringr::str_replace_all(str, unlist(as.list.environment(tokens)))
```
[the remark this answers is not in the document]
Comment (alandipert) [reply R/shiny-recorder.R]: Nope, no reason - will fix.

**Pull request merged**